proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,112 @@
+ will fall back to the existing SDK resolution strategy: `%PATH%` followed by machine wide installations.
+
+This lookup will stop on the first match which means it won't necessarily find the best match. Consider a scenario with a global.json that has:
+
+```json
+{
+  "sdk": {
+    "additionalPaths": [ ".dotnet" ],
+    "version": "7.0.200",
+    "rollForward": "latestFeature"
+  }
+}
+```
+
+In a scenario where the `.dotnet` directory had 7.0.200 SDK but there was a machine wide install of 7.0.300 SDK, the host resolver would pick 7.0.200 out of `.dotnet`. That location is considered first, it has a matching SDK and hence discovery stops there.
+
+The values in the `additionalPaths` property can be a relative or absolute path. When a relative path is used it will be resolved relative to the location of global.json. These values also support the following substitutions:
+
+- `"$user$"`: this matches the user-local installation point of .NET SDK for the current operating system: `%LocalAppData%\Microsoft\dotnet` on Windows and `$HOME/.dotnet` on Linux/macOS.
+- `"$machine$"`: this matches the machine installation point of .NET for the current operating system.
+- `%VARIABLE%/$VARIABLE`: environment variables will be substituted. Either the Windows or Unix format can be used here and will be normalized for the operating system the host resolver executes on.
+
+This design requires us to only change the host resolver. That means other tooling like Visual Studio, VS Code, MSBuild, etc ... would transparently benefit from this change. Repositories could update global.json to have `additionalPaths` support `.dotnet` and Visual Studio would automatically find it without any design changes.
+
+## Considerations
+### Installation Points
+One item to keep in mind when considering this area is the .NET SDK can be installed in many locations. The most common are:
+
+- Machine wide
+- User wide: `%LocalAppData%\Microsoft\dotnet` on Windows and `$HOME/.dotnet` on Linux/macOS.
+- Repo: `.dotnet`
+
+Our installation tooling tends to avoid redundant installations. For example, if restoring a repository that requires 7.0.400, the tooling will not install it locally if 7.0.400 is installed machine wide. It also will not necessarily delete the local `.dotnet` folder or the user wide folder. That means developers end up with .NET SDK installs in all three locations but only the machine wide install has the correct ones.
+
+As a result solutions like "just use .dotnet, if it exists" fall short. It will work in a lot of casse but will fail in more complex scenarios. To completely close the disconnect here we need to consider all the possible locations.
+
+### Do we need additionalPathsOnly?
```

Comment:
Did you consider that having the `$machine$` means that:

```json
"additionalPaths": [ ".dotnet", "$machine$" ],
"additionalPathsOnly": "true"
```

is equivalent to
```json
"additionalPaths": [ ".dotnet" ],
"additionalPathsOnly": "false"
```

Would it make sense to only use `additionalPaths` - and maybe rename it to something like `installPaths`. And if specified it would the only list searched?

Reply:
I did think about that and I agree it's a valid approach to this. I've frankly waffled on which I think is the better approach here. 

I think in most scenarios customers would want to consider machine based installations. No data here, just intuition. The `$machine$` value is a bit of a _magic_ value. In the `installPaths` approach that would mean that `$machine$` would end up in the majority of uses of `installPaths`. Had a hard time convincing myself that was the best outcome which is why I ended up with the current approach.

Reply:
I'm not advocating for either way, but having `$machine$` is slightly more flexible, because you can switch the order:

```json
"additionalPaths": [ "$machine$", ".dotnet" ],
```

Not sure why anyone would want to do that, but I'm sure someone will come up with a use-case at some point.

Reply:
> Not sure why anyone would want to do that, but I'm sure someone will come up with a use-case at some point.

That's basically what the Arcade script does that acquires the .NET SDK. It first checks if a direct match (based on the sdk version) is found in the %PATH% location and only if not then downloads the SDK to the local ".dotnet" directory.

That script could honor the selection here to make that behavior more configurable.

Reply:
Nit: `$machine$` doesn't mean anything to me. I'd call this `$default_install$` or similar.

And yeah, if we have a way to refer to the default install I would change `additionalPaths` to `installPaths` and just search those paths precisely in order.

Reply:
Actually, we still need to know whether to check `$PATH`.

Reply:
I'd prefer not having a separate `additionalPathsOnly` property.

Given that the magic is much less in 7+ (we only look for the SDK relative to the running `dotnet` and I don't recall fallout from it), I'd expect `$machine$`/`$default_install$` shouldn't end up in most uses.

Since the fallback logic would be to look relative to the running `dotnet`, I'd still consider that only local deployment and I could see always looking next to running `dotnet` (after `additionalPaths`/`installPaths`) being reasonable.